xprocess.py

```diff
@@ -47,17 +47,20 @@ def terminate(self, **kwargs):
         if not self.pid or not self.isrunning():
             return 0
         timeout = 20
-        parent = psutil.Process(self.pid)
-        kill_list = [parent]
-        if kill_proc_tree:
-            kill_list += parent.children(recursive=True)
-        for p in kill_list:
-            p.send_signal(signal.SIGTERM)
-        _, alive = psutil.wait_procs(kill_list, timeout=timeout)
-        for p in alive:
-            p.send_signal(signal.SIGKILL)
-        _, alive = psutil.wait_procs(kill_list, timeout=timeout)
-        if alive:
+        try:
+            parent = psutil.Process(self.pid)
```

Comment:
Looks good!

To be ultra-safe, how about moving this line outside this try/finally, into a separate try/except:

```python
try:
    parent = psutil.Process(self.pid)
except NoSuchProcess:
    return 0
```

This is a bit more "beg for forgiveness" than "ask for permission" approach taken in line 47. That line can then just be changed to:

```python
if not self.pid:
    return 0
```

WDYT?

Reply:
Nice catch. It is (as you said) unlikely to happen, but definitely possible if the process is externally terminated, for example, due to some exceptional circumstance like extreme resource starvation where it would get forcefully terminated by the kernel.